ang/test/civicase/case/services/add-case.service.spec.js

```diff
@@ -0,0 +1,133 @@
+/* eslint-env jasmine */
+
+(($, loadForm, getCrmUrl) => {
+  describe('AddCaseService', () => {
+    let $window, AddCase, mockedFormPopUp, CaseCategoryWebformSettings;
+
+    beforeEach(() => {
+      $window = { location: { href: '' } };
+    });
+
+    describe('Button Visibility', () => {
+      let isButtonVisible;
+
+      beforeEach(module('civicase-base', 'civicase'));
+      beforeEach(() => {
+        injectDependencies();
+      });
+
+      describe('when the user can add new cases', () => {
+        beforeEach(() => {
+          CRM.checkPerm.and.returnValue(true);
+
+          isButtonVisible = AddCase.isVisible();
+        });
+
+        it('displays the button', () => {
+          expect(isButtonVisible).toBe(true);
+        });
+      });
+
+      describe('when the user cannot add new cases', () => {
+        beforeEach(() => {
+          CRM.checkPerm.and.returnValue(false);
+
+          isButtonVisible = AddCase.isVisible();
+        });
+
+        it('does not display the button', () => {
+          expect(isButtonVisible).toBe(false);
+        });
+      });
+    });
+
+    describe('click handler', () => {
+      beforeEach(module('civicase', ($provide) => {
+        CaseCategoryWebformSettings = jasmine.createSpyObj('CaseCategoryWebformSettings', ['getSettingsFor']);
+        $provide.value('CaseCategoryWebformSettings', CaseCategoryWebformSettings);
+      }));
+
+      describe('when the new case web form url configuration value is defined', () => {
+        beforeEach(module('civicase', ($provide) => {
+          $provide.value('$window', $window);
+          CaseCategoryWebformSettings.getSettingsFor.and.returnValue({
+            newCaseWebformUrl: '/someurl',
+            newCaseWebformClient: 'cid'
+          });
+        }));
+
+        beforeEach(() => {
+          injectDependencies();
+          AddCase.clickHandler('case', '5');
+        });
+
+        it('redirects the user to the configured web form url value', () => {
+          expect($window.location.href).toBe('/someurl?cid=5');
+        });
+      });
+
+      describe('when the new case web form url configuration value is not defined', () => {
+        beforeEach(module('civicase-base', 'civicase', ($provide) => {
+          CaseCategoryWebformSettings.getSettingsFor.and.returnValue({ newCaseWebformUrl: null });
+          $provide.value('$window', $window);
+        }));
+
+        beforeEach(() => {
+          injectDependencies();
+          mockFormPopUpDom();
```

Comment:
Can we also add the dependencies injection and mocking to the top of the spec file?

Reply:
The mock can still be called from a function as this, but at the top of the file.